File: frontend/src/domain/models/ExerciseMedia.ts
import type { MediaKind } from '../enums/MediaKind'
import type { MediaSource } from '../enums/MediaSource'

/**
 * One media item attached to an exercise. Immutable.
 *
 * Abstracts over images and videos so the UI renders from a single list and
 * real videos (a curated YouTube mapping, or coach uploads in T1) drop in as a
 * pure data change — no UI rewrite. Today the static dataset yields only
 * `Image`/`File` items; the video paths are ready and dormant.
 */
export interface ExerciseMedia {
  readonly kind: MediaKind
  readonly source: MediaSource
  /** Locator: an absolute URL for `File`, or the video id for `YouTube`. */
  readonly url: string
  /** Optional still to show as a thumbnail/poster (e.g. a video cover frame). */
  readonly thumbnailUrl?: string
}

File: backend/src/main/java/com/musclemap/catalog/ExerciseMedia.java
package com.musclemap.catalog;

import jakarta.persistence.Column;
import jakarta.persistence.Embeddable;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

/**
 * One media item (image or video) attached to an {@link Exercise}. For
 * {@code FILE} media the {@code url} is an absolute address; for {@code YOUTUBE}
 * it is the video id. Owned by the exercise aggregate (table {@code exercise_media}).
 */
@Embeddable
@Getter
@Setter
@NoArgsConstructor
public class ExerciseMedia {

    @Enumerated(EnumType.STRING)
    @Column(name = "kind", length = 10, nullable = false)
    private MediaKind kind;

    @Enumerated(EnumType.STRING)
    @Column(name = "source", length = 10, nullable = false)
    private MediaSource source;

    @Column(name = "url", length = 500, nullable = false)
    private String url;

    @Column(name = "thumbnail_url", length = 500)
    private String thumbnailUrl;

    public ExerciseMedia(MediaKind kind, MediaSource source, String url, String thumbnailUrl) {
        this.kind = kind;
        this.source = source;
        this.url = url;
        this.thumbnailUrl = thumbnailUrl;
    }
}
